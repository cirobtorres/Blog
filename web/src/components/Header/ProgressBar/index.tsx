"use client";

import { useEffect } from "react";

const ProgressBar = ({ documentId }: { documentId: string }) => {
  const linearProgressBarOnScroll = () => {
    const elementHeight = document.getElementById(documentId)?.scrollHeight;
    const progressBar = document.getElementById("progress-bar-blur");
    const progressBarBlur = document.getElementById("progress-bar");
    const scrollTop = window.scrollY;
    if (elementHeight && progressBar && progressBarBlur) {
      const percentage =
        scrollTop < elementHeight ? (scrollTop / elementHeight) * 100 : 100;
      progressBarBlur.style.width = `${percentage}%`;
      progressBar.style.width = `${percentage}%`;
    }
  };

  useEffect(() => {
    window.addEventListener("scroll", linearProgressBarOnScroll);
    // cleanup function
    return () => {
      window.removeEventListener("scroll", linearProgressBarOnScroll);
    };
  }, []);

  return (
    <div className="fixed top-[calc(100%_+_1px)] left-0 h-1 w-full inline-grid">
      <div
        id="progress-bar-blur"
        className="w-0 h-full col-start-1 row-start-1 bg-gradient-to-r from-transparent to-blog-foreground-highlight blur-xl hidden max-lg:block"
      />
      <div
        id="progress-bar"
        className="w-0 h-full col-start-1 row-start-1 bg-gradient-to-r from-transparent to-blog-foreground-highlight hidden max-lg:block"
      />
    </div>
  );
};

export default ProgressBar;
