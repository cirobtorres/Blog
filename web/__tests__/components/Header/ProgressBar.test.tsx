import ProgressBar from "../../../src/components/Header/ProgressBar";
import { render, screen, fireEvent } from "@testing-library/react";

describe("ProgressBar", () => {
  const pageHeight = 5000;

  // beforeEach(() => {});

  afterAll(() => {
    jest.clearAllMocks();
  });

  it("renders bar component with starting width of 0%", () => {
    render(<ProgressBar />);
    const progressBar = screen.getByTestId("progress-bar");
    const progressBarBlur = screen.getByTestId("progress-bar-blur");
    expect(progressBar.style.width).toBe("0%");
    expect(progressBarBlur.style.width).toBe("0%");
  });

  it("renders bar component with accordingly y coordinate scroll", () => {
    render(<ProgressBar />);

    jest
      .spyOn(document.documentElement, "scrollHeight", "get")
      .mockImplementation(() => pageHeight); // Define a page height

    fireEvent.scroll(window, { target: { scrollY: pageHeight / 2 } }); // Scroll below 50% of that page height at least

    const progressBar = screen.getByTestId("progress-bar");
    const progressBarBlur = screen.getByTestId("progress-bar-blur");

    const progressBarPercentage = progressBar.style.width.match(/\d+/); // It is returning a string of a float like: "59.07372400756144%"
    const progressBarBlurPercentage = progressBarBlur.style.width.match(/\d+/); // It is returning a string of a float like: "59.07372400756144%"

    const progressNumber = Number(
      progressBarPercentage && progressBarPercentage.length > 0
        ? progressBarPercentage[0]
        : 0
    );
    const progressBlurNumber = Number(
      progressBarBlurPercentage && progressBarBlurPercentage.length > 0
        ? progressBarBlurPercentage[0]
        : 0
    );

    expect(progressNumber).toBeGreaterThanOrEqual(50);
    expect(progressBlurNumber).toBeGreaterThanOrEqual(50);
    expect(progressNumber).toEqual(progressBlurNumber);
  });

  it("applies and removes event listener when component mounts and unmounts respectively", () => {
    const addEventListenerSpy = jest.spyOn(window, "addEventListener");
    const removeEventListenerSpy = jest.spyOn(window, "removeEventListener");
    const { unmount } = render(<ProgressBar />);
    expect(addEventListenerSpy).toHaveBeenCalledWith(
      "scroll",
      expect.any(Function)
    ); // Checks if scroll event was added
    removeEventListenerSpy.mockClear(); // Clear component and checks if scroll event was removed
    unmount(); // Unmounting component
    expect(removeEventListenerSpy).toHaveBeenCalledWith(
      "scroll",
      expect.any(Function)
    );
  });

  it("renders progress bar with accessibility attributes and classes", () => {
    render(<ProgressBar />);
    const progressDiv = screen.getByTestId("progress");
    expect(progressDiv).toHaveAttribute("role", "progressbar");
    expect(progressDiv).toHaveAttribute("aria-live", "polite");
    expect(progressDiv).toHaveAttribute("aria-labelledby", "progressbar-label");
    expect(progressDiv).toHaveClass(
      "fixed top-[calc(100%_+_1px)] left-0 h-1 w-full inline-grid"
    );

    const progressBar = screen.getByTestId("progress-bar");
    const progressBarBlur = screen.getByTestId("progress-bar-blur");

    // Mobile only
    expect(progressBar).toHaveClass("hidden max-lg:block");
    expect(progressBarBlur).toHaveClass("hidden max-lg:block");

    // Both bars must overlap each other
    expect(progressBar).toHaveClass("col-start-1 row-start-1");
    expect(progressBarBlur).toHaveClass("col-start-1 row-start-1");

    // Progress bar blur must keep any blur class
    expect(progressBarBlur).toHaveClass(/.*blur.*/);
  });

  it("matches the snapshot", () => {
    const { asFragment } = render(<ProgressBar />);
    expect(asFragment()).toMatchSnapshot();
  });
});
