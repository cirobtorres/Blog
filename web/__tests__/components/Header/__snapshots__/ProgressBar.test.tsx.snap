// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar matches the snapshot 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="progressbar-label"
    aria-live="polite"
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="0"
    class="fixed top-[calc(100%_+_1px)] left-0 h-1 w-full inline-grid"
    data-testid="progress"
    id="progress"
    role="progressbar"
  >
    <div
      class="h-full col-start-1 row-start-1 bg-gradient-to-r from-transparent to-blog-foreground-highlight blur-xl rounded-full hidden max-lg:block"
      data-testid="progress-bar-blur"
      id="progress-bar-blur"
      style="width: 0%;"
    />
    <div
      class="h-full col-start-1 row-start-1 bg-gradient-to-r from-transparent to-blog-foreground-highlight rounded-full hidden max-lg:block"
      data-testid="progress-bar"
      id="progress-bar"
      style="width: 0%;"
    />
    <span
      class="sr-only"
      id="progressbar-label"
    >
      Progresso de rolagem da p√°gina
    </span>
  </div>
</DocumentFragment>
`;
